### ISSUE TITLE
API call latencies regressed due to upgrade to golang 1.12

### ISSUE THREAD

**Issue by mborsz** [kubernetes/kubernetes#75833]
Something has changed between 320 and 321 runs and the request latency has increased significantly.

The graphs below shows request latency (non watch) over time for a single test run:

[320](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/320):
![FYrJr3W9WRY](https://user-images.githubusercontent.com/2559168/55164808-a189ef00-516c-11e9-97fa-931d507fbcfe.png)

[321](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/321):
![vHedLTUA8zO](https://user-images.githubusercontent.com/2559168/55167490-5faf7780-5171-11e9-97b3-501e7e3b4309.png)

In my opinion they looks significantly different: in 320 most of the spikes are lower than 0.8s while 321 most of the spikes exceeds 1.6s.

I checked few runs below 320 and have latency pattern similar to 320 and few runs after 321 and they have similar pattern to the 321.

We are seeing similar latency growth in kubemark tests.

The most significant change between 321 and 320 was go 1.11.5 -> 1.12 migration.

/cc @wojtek-t @krzysied

**Comment by mborsz** [comment]
With @krzysied we are trying to prepare small scale repro that we can use for further debugging.

**Comment by liggitt** [comment]
#74778 was also in that range (cc @sttts @roycaihw for verification that nothing in that PR increased server load in non-alpha clusters)

**Comment by wojtek-t** [comment]
@krzysied did tests at head with reverting to go1.11 and metrics seemed better
He is working on confirming that and should paste results then, but it seems to be pretty clear that it's go regression.
We're also working on providing smaller-scale repro case to work closer with golang team (they used to be really helpful when debugging regressions in previous versions (IIRC 1.9 is where there were previous one).

**Comment by roycaihw** [comment]
I double checked #74778 and didn't find anything suspicious. The alpha function wasn't invoked in that window, and the PR is a no-op to non-alpha clusters

**Comment by krzysied** [comment]
Regarding what @wojtek-t wrote:
![kubemark_regression](https://user-images.githubusercontent.com/25119199/55189408-25100400-519e-11e9-9b6e-762f3896e61c.png)

I was mostly investigating the regression on kubemark-5000. Some of its latencies increased x3.

I run two tests: one with https://github.com/kubernetes/kubernetes/commit/7514c49ec12e7a8b6840142d723ec25b08a0917a checkout, and the other one with some commit checkout but without https://github.com/kubernetes/kubernetes/commit/36e934d0a9a004a5f6c0a06e22e89ec4fe583b6a (Update to use golang 1.12).

The 99th latency percentile of node listing was 3400.152ms in first case  and f 952.576ms in second one.

**Comment by wojtek-t** [comment]
Look into other metrics (I can't remember which exactly, delete pods? post pods? or sth like that) - which increased by almost 10x.

**Comment by wojtek-t** [comment]
This is the graph that I had on my mind, but yeah, it's more like 4x not 10x. But still quite huge drop:
![Screenshot from 2019-03-29 13-46-32](https://user-images.githubusercontent.com/10743879/55233585-5550b480-5229-11e9-941a-8a96b53e97f0.png)

**Comment by krzysied** [comment]
Reproducing the issue:

1. Checkout kubernetes repo - https://github.com/kubernetes/kubernetes.git.
- First broken version 36e934d0a9a004a5f6c0a06e22e89ec4fe583b6a (go1.12)
- Last working version 8c95a07b4bf179fd64a84215e041401a313ce0f6 (go1.11)

2. Checkout perf-tests repo - https://github.com/kubernetes/perf-tests.git.
 This repo is home for clusterloader2 - tool used for running the scalability tests.

3. In kubernetes directory run “make quick-release”. This will build kubernetes based on checkout version.

4. Create GCE cluster.

```
export KUBE_GCE_NETWORK=e2e-$USER
export KUBE_GCE_INSTANCE_PREFIX=e2e-$USER
export INSTANCE_PREFIX=e2e-$USER

export PROJECT_ID=<project_name>
export ZONE=<zone>

go run hack/e2e.go -- \
    --gcp-project=$PROJECT_ID \
    --gcp-zone=$ZONE \
    --check-leaked-resources \
    --gcp-master-size=n1-standard-8 \
    --gcp-node-image=gci \
    --gcp-node-size=n1-standard-8 \
    --gcp-nodes=16 \
    --provider=gce \
    --up \
    --test=false \
    --timeout=30m
```

5. Run kubemark test.
This step creates kubemark cluster (on top of the cluster created in step 4), run cluster loader test and tear down the cluster.

```
export KUBE_GCE_NETWORK=e2e-$USER
export KUBE_GCE_INSTANCE_PREFIX=e2e-$USER
export INSTANCE_PREFIX=e2e-$USER

export KUBEMARK_MASTER_COMPONENTS_QPS_LIMITS="--kube-api-qps=100 --kube-api-burst=100"

export PROJECT_ID=<project_name>
export ZONE=<zone>

go run hack/e2e.go -- \
    --gcp-project=$PROJECT_ID \
    --gcp-zone=$ZONE \
    --provider=gce \
    --kubemark \
    --kubemark-nodes=1000 \
    --test=false \
    --test-cmd=$GOPATH/src/k8s.io/perf-tests/run-e2e.sh \
    --test-cmd-args=cluster-loader2 \
    --test-cmd-args=--nodes=1000 \
    --test-cmd-args=--provider=kubemark \
    --test-cmd-args=--report-dir=/tmp/cluster_loader2/_artifacts \
    --test-cmd-args=--testconfig=testing/load/config.yaml \
    --test-cmd-args=--testoverrides=./testing/load/kubemark/500_nodes/override.yaml \
    --test-cmd-args="--enable-prometheus-server=false" \
    --test-cmd-name=ClusterLoaderV2 \
    --timeout=60m
```

The test run metrics will be exported to directory provided by the --report-dir flag. Inside this directory there will be APIResponsiveness_load_<time>.txt file containing 50, 90, and 90 percentiles of all apiserver call latencies. The latencies affected the most are:
 - Listing nodes
{
	"Resource": "nodes",
	"Scope": "cluster",
	"Subresource": "",
	"Verb": "LIST"
}
 - Pod deletion
{
	"Resource": "pods",
	"Scope": "namespace",
	"Subresource": "",
	"Verb": "DELETE"
}

- Listing services
{
	"Resource": "services",
	"Scope": "cluster",
	"Subresource": "",
	"Verb": "LIST"
}

I would suggest their 99th percentiles as the issue metric.

As for comparison:
-Go1.11 run:
  - Listing nodes - 162.409ms
  - Pod deletion - 41.302ms
  - Listing services - 28.277ms

-Go1.12 run:
  - Listing nodes - 293.724ms
  - Pod deletion - 56.2ms
  - Listing services - 46.183ms

6. Tear down GCE cluster

```
export KUBE_GCE_NETWORK=e2e-$USER
export KUBE_GCE_INSTANCE_PREFIX=e2e-$USER
export INSTANCE_PREFIX=e2e-$USER

export PROJECT_ID=<project_name>
export ZONE=<zone>

go run hack/e2e.go -- \
    --gcp-project=$PROJECT_ID \
    --gcp-zone=$ZONE \
    --check-leaked-resources \
    --gcp-master-size=n1-standard-8 \
    --gcp-node-image=gci \
    --gcp-node-size=n1-standard-8 \
    --gcp-nodes=16 \
    --provider=gce \
    --down \
    --test=false \
    --timeout=30m
```

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability-bugs

**Comment by wojtek-t** [comment]
We're trying to dig deeper into it to both root cause it to a smaller part of the system, but it is quite tricky.
Unfortunately, we don't have much more than guesses right now, though.

Some observations:
- I looked into kubemark-5000 runs into DELETE pods calls, and it seems that all of them for which we have traces (500ms+) - there are ~5000 of those in each of recent runs [previously 99th percentile was <200ms), there is a significant big of "Object deleted from database".
- That may suggest some etcd issues, but we don't see logs related to long-running-calls on etcd side. That may suggest some issues on the etcd client side, but we don't have anything concrete yet.

- It's also possible that golang upgrade changes something in serialization/deserialization - that's a second hypothesis (we've seen a bunch of traces of "Decoded delete options" (deserialization is also done in etcd client)), but I'm not convinced this is the case as of now

**Comment by wojtek-t** [comment]
We've spent significant time with golang folks and already have some findings.
To summarize:
- the microbenchmark from #76080 enabled Go team to find some regression on their side - the faulty commit from golang is: 07e738ec32025da458cdf968e4f991972471e6e9
- to cite Go folks: "To summarize the change: I removed the "fast" path for small page-level allocations (which currently always need to acquire the heap lock) in order to simplify code for subsequent changes. In effect, span allocations now all go through a balanced tree rather than just a linked list for small spans (I think it was all spans under 128 pages?)."

We aren't fully convinced that this is for sure the same thing that is causing regression in our scalability tests, but analyzing traces, profiles, golang tracesetc. suggests that problems are indeed around memory allocations or GC. So with high probabilty this is the same thing.

We are discussing with the way to prove it's the first thing and working with the to see what we can do with it.

**Comment by wojtek-t** [comment]
It appeared that the commit 07e738ec32025da458cdf968e4f991972471e6e9 isn't the most faulty one for us - running scalability tests against that commit doesn't show significant drop in metrics.

In the lack of better ideas, we decided to bisect changes in golang by running full scalability suite on 5k-node cluster against them. This will take couple more days though...

**Comment by wojtek-t** [comment]
Just to update this thread - we're bisecting Go 1.12 branch and there aren't that many remaining commits.
We already know that:
9d23975d89e6cc3df4f2156b2ae0df5d2cef16fb is a good one
10faf001077ea6b3907864f16a9f9099a7ba93 is a bad one

I hope that we will know the exact faulty commit on Monday.

**Comment by dims** [comment]
Looking through https://github.com/golang/go/compare/9d23975d89e6cc3df4f2156b2ae0df5d2cef16fb...10faf001077ea6b3907864f16a9f9099a7ba93

the following stand out for me
- https://github.com/golang/go/commit/faf187fb8e2ca074711ed254c72ffbaed4383c64
- https://github.com/golang/go/commit/8e093e7a1cd8a092f23717cb8f34bca489a3eee5

/me places bets on one of those 2 :)

**Comment by wojtek-t** [comment]
That is our bet too - we're working with Go folks to confirm and fix/mitigate this.

**Comment by wojtek-t** [comment]
I'm sorry for delay with update here - the current update is as following:
- after the next round of bisections, we know that:
ea27cd35fa07b874b3dff5d2f83dd401a361865e  is a good one
6f4dc1ccf9735013fdb7cd044bda29d19bebb906  is a bad one

During my discussions with golang folks, the outcome was that we believe that
8e093e7a1cd8a092f23717cb8f34bca489a3eee5 is the faulty one, as it adds a bunch of syscall to second-slowest allocation path
faf187fb8e2ca074711ed254c72ffbaed4383c64 (built on top of the previous one) should only help

So the next step was to verify
8e093e7a1cd8a092f23717cb8f34bca489a3eee5
and indeed it is a bad one (and actually as expected, the metrics were worse than in the so far confirmed oldest bad commit: 6f4dc1ccf9735013fdb7cd044bda29d19bebb906 )

This is what we already know.

What we're doing now is that golang folks provided us two patches that they think should mitigate the issue and I'm trying to confirm if that is really true. We should have results of this one later today.
The two potential outcomes are:
- it solves our issue; then I will be working with golang folks to patch 1.12 with those
- it doesn't solve our issue: then I will be working on final confirmation that it's exactly this commit (by checking the previous one, and the next one to confirm it really helps) and working further with golang folks on what else we can do to mitigate it (and also I will try to understand deeper what exactly is happening there, I just have some rough understanding now).

**Comment by dims** [comment]
Very very cool! thanks @wojtek-t

**Comment by dims** [comment]
cc @liggitt

**Comment by wojtek-t** [comment]
Just to give an update on that.

- we fully confirmed that the regression is due to 8e093e7a1cd8a092f23717cb8f34bca489a3eee5 (using go1.12.1 with manually reverted that commit reverts the behavior)
- however, according to golang fokls that commit was fixing some things, and there are couple commits built on top of them after 1.12, so it's not-trivial to revert it (we are still looking for other options)

- the patches that golang team provided helps, but don't fix the problem (they roughly reduce the problem to 50%)
- yesterday I got another two patches that may help and we're validating them now

- in the meantime we're also working with golang folks to try gathering more useful data for them for debugging - they already have some patches to apply in golang and I'm working on piping them through k8s and our build scripts - I hope I will have them ready today for an over-night run

**Comment by wojtek-t** [comment]
To give an update on that - the current state of it is as following:
- that particular commit can't be just reverted, as it was fixing some issue
- however, golang folks have a PR that is mitigating that issue in a different way
- we have verified with custom build of that fully mitigates regression on our side

The current plan is as following:
- https://github.com/golang/go/commit/8c05d67661c966f5130e51ca685b0c70a5a929ff has been merged to head of Go [that is the mitigation commit]
- it will be cherrypicked to 1.12 branch
- additionally in 1.12 we will revert the offending commit: 8e093e7a1cd8a092f23717cb8f34bca489a3eee5
We expect a new patch release of Go with those two commits to be released on May 1st.

@mknyszek is currently my main point-of-contact coordinating that on golang side.

**Comment by dims** [comment]
thanks for the update @wojtek-t ! we will still be in 1.15 alpha* phases so that's great news.

**Comment by wojtek-t** [comment]
@dims - the trick is that go 1.12 is also in 1.14, so we should also cherrypick 1.14 branch - currently there is visible regression in 1.14

**Comment by dims** [comment]
@wojtek-t agree. i'd support that

**Comment by wojtek-t** [comment]
The commits that we were waiting for are already there:
https://github.com/golang/go/commit/fd3676302e8436ec9243ae0695582bbb3f31dc3a
https://github.com/golang/go/commit/438b1a5dae01812a2b91b648c762414d4854ccd7

Now we are waiting for next 1.12 patch release of Go to update Kubernetes to use, the ETA is May 1st.

@dims - given that I will be on vacation for the rest of this week, would you be able to help with that?

**Comment by dims** [comment]
@wojtek-t yes i can, please let me know what needs to be done (so i can figure out if i have access before you head out)

**Comment by wojtek-t** [comment]
@dims  - thanks a lot. I didn't do upgrade of go for 2 years or so, so I'm not entirely sure how (if?) the process changed in the meantime. @cblecker - was doing the last one: https://github.com/kubernetes/kubernetes/pull/76576, so I'm sure he will be able to help with that.

**Comment by dims** [comment]
Ah just update to latest golang when it gets out. yes, i can do that. i was wondering if you wanted me to kick off some CI runs or something. 

Have a good vacation @wojtek-t

**Comment by wojtek-t** [comment]
@dims - thanks a lot!
Yes - it wasn't clear, but I was testing that and it helps us, so no need to run another tests (we will get final confirmation from postsubmits).

**Comment by cblecker** [comment]
I can also help out to get this upgrade in when it’s released. Thanks for all your work on this @wojtek-t!

**Comment by cblecker** [comment]
Looks like 1.12.5 went out this afternoon. PR open (https://github.com/kubernetes/kubernetes/pull/77528), but expect tests to fail:
- Waiting for the docker hub image to publish. The PR is merged, so this should happen shortly.
- Need a new version of the bazel rules_go to be published.

**Comment by wojtek-t** [comment]
Just to clarify the above (from @cblecker :
```
oh, so I saw the auto update ran here: https://github.com/docker-library/golang/commits/master

but it looks like the update for the tags hasn't merged yet: https://github.com/docker-library/official-images/pull/5867
```

**Comment by dims** [comment]
Yea, it takes some time for the images to get published.

**Comment by cblecker** [comment]
```
Digest: sha256:e52f9c9ea80fa182a78bd83a9f9e107d173b8e7fd922dbb0c60c8020a88fb7f6
Status: Downloaded newer image for golang:1.12.5
```

Looks like new images are live.

**Comment by cblecker** [comment]
Reopening this so we can verify that everything is back to normal in CI

**Comment by wojtek-t** [comment]
For me it's pretty clear that this was resolved:

http://perf-dash.k8s.io/#/?jobname=kubemark-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness&Resource=pods&Scope=namespace&Subresource=&Verb=DELETE

http://perf-dash.k8s.io/#/?jobname=kubemark-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness&Resource=nodes&Scope=cluster&Subresource=&Verb=LIST


Now there is one more thing we should do, which is cherrypick it to 1.14 branch.
@cblecker @dims ^^

**Comment by cblecker** [comment]
Opened pick: https://github.com/kubernetes/kubernetes/pull/77831

**Comment by tpepper** [comment]
@wojtek-t has SIG Scalability done any one-off custom scale/perf testing with the new golang patch release and the k8s 1.14 branch content?

Do you see this cherry pick as a 1.14.2 release blocker?  Or would a week or two of focused testing with updated golang in the 1.14 k8s branch and a 1.14.3 release by end of May be a sufficiently safe yet expedient response?

**Comment by wojtek-t** [comment]
> @wojtek-t has SIG Scalability done any one-off custom scale/perf testing with the new golang patch release and the k8s 1.14 branch content?

We were doing a bunch of tests to discover the real regression in golang 1.12 - the regression happened within 1.14 release timeframe and the bisection was done on that branch.
So yes - I'm 100% confident that it helps.

> Do you see this cherry pick as a 1.14.2 release blocker? Or would a week or two of focused testing with updated golang in the 1.14 k8s branch and a 1.14.3 release by end of May be a sufficiently safe yet expedient response?

Tough question - I would love to see it in 1.14.2, otoh given how long it took to debug and fix it, I can see release team pushing it to 1.14.3. But if we could have that in 1.14.2, it would be extremely helpful.

**Comment by hoegaarden** [comment]
#77831 just merged and we let 1.14 soak for one or two days before we cut `v1.14.2`.

Let's keep this issue open until we have some data on the latencies after that merge.

SIG Scalability, can you please provide some data / graphs about the latency in a (couple of) days for the `release-1.14` branch?

**Comment by tpepper** [comment]
@wojtek-t is there any evidence we'll be able to observe in https://testgrid.k8s.io/sig-release-1.14-blocking for validation?  Or is it possible for SIG Scalability to trigger any custom runs now that the cherry-pick has merged to the 1.14 branch?

**Comment by wojtek-t** [comment]
We aren't running large tests on release branches due to cost of them...
We don't have any free quota to run it now, due to huge number of regressions that we have at head that we're currently debugging.

But I'm 100% confident with this change. What I was personally doing in the past:
- I was running tests against commit when we bumped to 1.12.0 with 1.12.5 (which was from beginning of March, so still in 1.14 timeline) and the effect was exactly the same as the effect against head we got after merging it.
- I also understand changes in Go between 1.12.4 and 1.12.5 and their effect on performance
- And we also understand this regression, and why the fixes help, which has nothing to do with any changes that are happening to kubernetes.

So I'm entirely sure that if you didn't observe any correctness issues, from performance perspective it is good-to-go.
